<a href="https://colab.research.google.com/github/yesong98/Exploration/blob/main/Exploration_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. LORD_DIGITS(예측)

In [78]:
import sklearn

print(sklearn.__version__)

1.0.2


(1)필요한 모듈 import

In [79]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

(2)데이터 준비

In [80]:
digits = load_digits()

(3)데이터 이해하기

In [81]:
#Feature Data 지정하기
digits_data = digits.data

In [82]:
#Label Data 지정하기
digits_label = digits.target

In [83]:
#Target Names 출력해 보기
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [84]:
#데이터 Describe 해 보기
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

(4) train, test 데이터 분리

In [85]:
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=10)

(5)모델 학습 및 예측

In [86]:

import pandas as pd

print(pd.__version__)

1.3.5


In [87]:
#자료형으로 변환
digits_df = pd.DataFrame(data=digits_data, columns=digits.feature_names)
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [88]:
digits_df["label"] = digits.target
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

In [89]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
print(decision_tree._estimator_type)

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.92      0.96        37
           1       0.83      0.88      0.86        34
           2       0.85      0.82      0.84        34
           3       0.80      0.90      0.85        40
           4       0.76      0.76      0.76        34
           5       0.91      0.91      0.91        32
           6       0.89      0.89      0.89        37
           7       0.95      0.90      0.92        40
           8       0.77      0.73      0.75        33
           9       0.85      0.87      0.86        39

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.86       360



# 1) Decision Tree를 이용한 예측

* decision tree에서 0이라는 숫자인식을 먼저 보면 precision의 수치가1.00이므로 숫자를 기입했을 때 오류가 날 확률이 없다는 것을 알 수 있다. 반면 0으로 확실하게 예측하는 수치는 0.92다. 아직 확실하게 예측하지 못하는 정도가 0.08인것이다.
* 숫자 1의 경우 오류가 날 확률은 0.17정도로 숫자 0에 비해서는 오류가 날 확률이 높다는것을 알 수 있고 recall의 값을 통해 확실히 인식하는 수치가 0.88인것을 알 수 있다.  
* 이렇게 전체적으로 숫자 이미지에 대한 오류를 확인해보면 가장 오류가 적은 것은 숫자0이고 가장 오류가 많이 일어나는 것은 숫자 4이다. 숫자 4의 precision의 수치가 10개의 숫자 중 가장 낮기 때문이다. 하지만 숫자를 확실하게 인식하는 수치는 숫자 8의 recall 값이 0.73으로 가장 낮은 수치를 보이고 있다. 
* 숫자를 인식하는데 있어 FP의 수치가 높은 것은 숫자 4이고 FN의 수치가 높아 정확도가 떨어지는 것은 숫자8인 것이다. 


In [90]:
#Random forest
from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_label,
                                                    test_size=0.2,
                                                    random_state=21)
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.94      0.95        32
           1       0.97      1.00      0.99        36
           2       1.00      1.00      1.00        30
           3       0.98      0.98      0.98        41
           4       0.94      0.97      0.95        32
           5       1.00      0.98      0.99        46
           6       1.00      0.97      0.98        32
           7       0.98      1.00      0.99        40
           8       0.93      0.98      0.95        42
           9       1.00      0.93      0.96        29

    accuracy                           0.97       360
   macro avg       0.98      0.97      0.97       360
weighted avg       0.98      0.97      0.98       360



# 2) Random Forest를 이용한 예측

* Random Forest는 Decision Tree의 단점을 보완한 방법으로 여러가지 의견을 통합하는 "앙상블" 기법을 이용한다. 의사 결정 트리를 만드는데 있어 쓰이는 요소들을 무작위적으로 선정하는 것이다.  
* 이 표에서는 precision이 1.00의 수치를 보인 것이 숫자 2,5,6,9가 있고 Decision Tree보다 숫자 예측이 높다는 것을 알 수 있다. 
* 이 중에서도 숫자 2는 precision과 recall의 수치가 모두 1.00이므로 가장 잘 예측할 수 있는 숫자인 것이다. 
* 위의 Decision Tree와 마찬가지로 숫자 4의 precison 수치가 낮은 것을 알 수 있다. 숫자 4의 표기나 모양이 가장 인식하기 어렵거나 브러쉬에 따른 형태가 너무 다양해서 다른 숫자들에 비해 오류의 확률이 높지 않은가 생각한다.
* 또 Decision Tree와 다르게 숫자 9가 prcesion은 1.00으로 높게 나왔으나 recall은 0.93으로 가장 낮은 recall값을 보인다. 이는 오류가 날 확률은 없어보이지만 확실하게 인식하는 정도가 다른 숫자에 비해 낮은 수치를 보인다.

In [91]:
#SVM모델 사용 
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.95      1.00      0.97        36
           2       1.00      1.00      1.00        30
           3       1.00      1.00      1.00        41
           4       0.97      1.00      0.98        32
           5       0.98      1.00      0.99        46
           6       1.00      1.00      1.00        32
           7       1.00      1.00      1.00        40
           8       0.98      0.95      0.96        42
           9       1.00      0.93      0.96        29

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



# 3) SVM모델을 이용한 예측

* 전체적으로 수치를 보면 위에서의 Decision Tree와 Random Forest보다 오류가 적은 것을 알 수 있다.
* precision의 수치가 1.00인 것은 숫자 0,2,3,6,7,9가 있고 recall의 경우 수치가 1.00인 것이 숫자 1,2,3,4,5,6,7이 있다.
* precision과 recall의 값이 1.00인 것은 2,3,6,7 이다.(오류가 나지 않았다)
* SVM모델을 이용한 결과 숫자 1이 오류가 날 확률이 0.05정도 존재하고 숫자 9의 예측이 가장 부정확 했다는 것을 알 수 있다.

In [92]:
#SGD Classifier모델 사용
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.72      1.00      0.84        36
           2       1.00      1.00      1.00        30
           3       1.00      0.78      0.88        41
           4       1.00      0.91      0.95        32
           5       0.94      0.98      0.96        46
           6       1.00      0.94      0.97        32
           7       1.00      0.93      0.96        40
           8       0.80      0.93      0.86        42
           9       0.88      0.72      0.79        29

    accuracy                           0.92       360
   macro avg       0.93      0.91      0.92       360
weighted avg       0.93      0.92      0.92       360



# 4) SGD모델을 이용한 예측

* SVM 모델에 비해 에측 정도가 낮은 걸 알 수 있다.
* precision의 수치가 1.00인 것은 숫자 0,2,3,6,7,9가 있고 recall의 경우 수치가 1.00인 것이 숫자 1,2,3,4,5,6,7이 있다.
* precision과 recall의 값이 1.00인 것은 2,3,6,7 이다.(오류가 나지 않았다)
* SVM모델을 이용한 결과 숫자 1이 오류가 날 확률이 0.05정도 존재하고 숫자 9의 예측이 가장 부정확 했다는 것을 알 수 있다.

In [93]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        32
           1       0.95      0.97      0.96        36
           2       1.00      1.00      1.00        30
           3       0.98      1.00      0.99        41
           4       0.94      0.97      0.95        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       0.97      0.97      0.97        40
           8       1.00      0.95      0.98        42
           9       0.96      0.93      0.95        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


# 5) Logistic Regression를 이용한 예측
* 앞서 설명한 SGD모델에 비해서 적은 수치를 가진 숫자가 많이 보인다.
* precision과 recall의 값이 모두 1.00인 것은 숫자 0,2,6이 있다
* 가장 낮은 precision값을 보인 것은 0.94로 숫자 4이고 가장 낮은 recall값은 0.93으로 숫자9이다.

# 결론

* digit에서는 해당 숫자가 맞지 않는데 맞다고 판단하는 경우가 적어야 하므로 Precision이 중요하다. 
* Precision은 분모에 있는 FP가 낮을수록 커진다. Precision이 높아지려면 False Positive, 즉 음성인데 양성으로 판단하는 경우가 적어야한다. 
* 그러한 전제로 precision의 수치가 1.00이 많이 나온 SVM 모델이 적절하다고 생각되어지나 precision 수치가 1.00 만 많다고 해서 잘 예측했다 보긴 어렵다. 전체적인 precision 평균값을 비교해보기 위해서는 macro avg와 weighted avg의 값도 고려해야된다.
* 모델 중 precision의 macro avg와 weighted avg의 값이 가장 높은 모델은 0.99로  SVM모델이다.(평균값에 대한 순위: SVM > RandomForest = SGD = Logistic Regression > Decision Tree) 
* 그러므로 Digits 데이터를 이용한 예측에서 가장 적절한 모델 평가 지표는 SVM이라 생각한다.

# 2. LORD_WINE(예측)

(1)필요한 모듈 import 하기

In [94]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

(2)데이터 준비

In [95]:
wine = load_wine()

(3)데이터이해하기

In [96]:
wine_data = wine.data #Feature Data 지정하기
wine_label = wine.target #Label Data 지정하기

In [97]:
#Target Names 출력해 보기
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [98]:
#데이터 Describe 해 보기
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

(4)train, test 데이터 분리

In [99]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=21)


(5)다양한 모델로 학습시켜보기

In [100]:
#Decision Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      1.00      0.94        15
           1       0.80      0.80      0.80        10
           2       0.89      0.73      0.80        11

    accuracy                           0.86        36
   macro avg       0.86      0.84      0.85        36
weighted avg       0.86      0.86      0.86        36



# 1) Decision Tree를 이용한 예측

* precision은 대체적으로 0.8의 수치를 보이고 있고 recall은 0으로 분류한 수치가 가장 높게 나온다.
* precision의 평균은 0.86을, recall의 평균은 0.84를 가진다. 


In [101]:
#Random Forest
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00        11

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



# 2) Random Forest를 이용한 예측

* 모든 값이 1.00으로 오류 없이 와인을 잘 분류했다고 볼 수있다.


In [102]:
# SVM
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.80      0.83        15
           1       0.45      1.00      0.62        10
           2       0.00      0.00      0.00        11

    accuracy                           0.61        36
   macro avg       0.44      0.60      0.48        36
weighted avg       0.48      0.61      0.52        36



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# 3) SVM을 이용한 예측

* class 0으로 분류하는데 있어서 큰 문제가 없었지만 class 2로 분류하는데 오류가 발생해 수치가 0.00으로 나온 것을 확인 할 수 있다.
* 다중분류(여러가지 class를 분류)하는데 있어서 SVM모델의 예측은 좋지 못한 수치를 가진다는 것을 알 수 있다.


In [103]:
# SGD Classifier
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.81      0.87      0.84        15
           1       0.83      0.50      0.62        10
           2       0.50      0.64      0.56        11

    accuracy                           0.69        36
   macro avg       0.72      0.67      0.67        36
weighted avg       0.72      0.69      0.69        36



In [104]:
#Logistic Regression
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00        11

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


# 결론

(1) 전제
* Wine 데이터는 와인의 종류를 구분하는 문제이다. 이의 경우 precision(정확도)를 통해 예측해보는 것이 적절하다. 
* 0, 1, 2로 분류된 class에서 0을 0으로 예측하고 1을 1로 2를 2로 예측을 잘 해주는 모델이 적합하기 때문이다. 
* 암을 양성과 음성을 분류하는 모델에서는 양성을 음성이라고 판단하면 안되는 상황이 있지만 wine은 그런 상황이 딱히 없다. 0을 1이라 예측하면 잘못분류했을 뿐 특별한 조건이 필요한 것은 없다.

(2) 다중분류
* 왜 SVM모델은 앞서 예측한 digits에 비해 wine을 이용한 예측에서는 낮은 값이 도출되었을지 알아보니 다중분류보다 SVM모델은 이진분류에 더 적합하다는 것을 알 수 있었다.
* Wine과 같은 다중분류문제에서는 Random Forest와 Logistic Regression이 가장 적절한 모델이라 생각한다.
* Random Forest에서 의사결정 나무를 다수 생성하고 '앙상블'을 통해 class를 분류하는 것이 wine데이터에서는 긍정적인 성능을 보이기 때문이다.
* 또한 Logistic Regression의 경우 구분해야할 클래스가 3개 이상일 때 우리는 다중 분류 로지스틱 회귀분석을 이용할 수 있다. 이진 분류 때와 다른 점은 확률로 변환시켜주는 함수를 시그모이드 함수가 아니라 소프트맥스 함수를 사용한다는 것이다.
([출처] 다중 분류 로지스틱 회귀분석(feat.소프트맥스)|작성자 생각하는 산티아고)

(3)정확도
* 무엇보다 모델 성능을 종합적으로 평가하려면 precision과 recall을 모두 조사해야 한다. F1 점수는 두 가지를 모두 고려하는 유용한 지표 역할을 한다. 
* F1 점수의 accuary를 비교해 정확도를 알아보았다.
* accuary 는 모델을 진행한 순서대로 0.86, 1.00, 0.61, 0.61, 1.00의 정확도를 보이고 있는 것을 알 수 있다

(4)적절한 모델
* Random Forest, Logistic Regression 모델



# 3. load_breast_cancer: 유방암 여부를 진단

(1) 필요한 모듈 import하기

In [105]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

(2) 데이터 준비

In [106]:
breast_cancer = load_breast_cancer()

(3)데이터 이해하기

In [107]:
breast_cancer_data = breast_cancer.data #Feature Data 지정하기
breast_cancer_label = breast_cancer.target #Label Data 지정하기

In [108]:
#Target Names 출력해 보기
breast_cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [109]:
#데이터 Describe 해 보기
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

(4)데이터 분리

In [110]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=10)

(5)다양한모델 학습시켜보기

In [111]:
#Decision Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.95      0.88        39
           1       0.97      0.89      0.93        75

    accuracy                           0.91       114
   macro avg       0.90      0.92      0.91       114
weighted avg       0.92      0.91      0.91       114



# 1) Decision Tree 모델을 이용한 예측
* 모델은 평균 90%의 precision(정밀도)와 92%의 recall(재현율)로 꽤 좋은 점수이다
* accuracy(정확도)가 91%인것을 알 수 있다


In [112]:
#Radom forest
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.95      1.00      0.97        39
           1       1.00      0.97      0.99        75

    accuracy                           0.98       114
   macro avg       0.98      0.99      0.98       114
weighted avg       0.98      0.98      0.98       114



# 2) Random Forest 모델을 이용한 예측
* 모델은 평균 98%의 precision(정밀도)와 99%의 recall(재현율)로 지금까지 가장 좋은 점수이다.
* accuracy(정확도)는 98%인것을 알 수 있다. 정확도도 상당히 높은 점수이다.


In [113]:
#SVM
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.82      0.88        39
           1       0.91      0.97      0.94        75

    accuracy                           0.92       114
   macro avg       0.93      0.90      0.91       114
weighted avg       0.92      0.92      0.92       114



# 3) SVM 모델을 이용한 예측
* 모델은 평균 93%의 precision(정밀도)와 90%의 recall(재현율)로 Decision Tree와 비슷한 수치를 가진다
* accuracy(정확도)가 92%인것을 알 수 있다


In [114]:
#SGD Classifier
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.69      0.81        39
           1       0.86      0.99      0.92        75

    accuracy                           0.89       114
   macro avg       0.91      0.84      0.86       114
weighted avg       0.90      0.89      0.88       114



# 4) SGD 모델을 이용한 예측
* 모델은 평균 92%의 precision(정밀도)와 85%의 recall(재현율)로 가장 낮은 recall값을 가진다.
* accuracy(정확도)는 89%인것을 알 수 있다

In [115]:
#Logitstic Regression
logistic_model.fit(X_train,y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.95      0.91        39
           1       0.97      0.93      0.95        75

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.93       114
weighted avg       0.94      0.94      0.94       114



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


# 5) Logistic Regression 모델을 이용한 예측
* 모델은 평균 93%의 precision(정밀도)와 94%의 recall(재현율)로 좋은 수치를 가진다
* accuracy(정확도)가 94%인것을 알 수 있다

# 결과
(1) 전제
* 암의 경우 양성인 것을 음성으로 판단하면 안되기 때문에 FN의 값이 줄어들어 Recall값이 높은 것이 적절한 결과라 할 수 있다.
* 이번 에측에서는 Recall값의 평균을 자세히 살펴 보아야한다.

(2) 분석
* recall(재현율)을 먼저 따져보았을 때 Randomforest(98%) > Logistic Regression(94%) > Decision Tree(92%) > SVM(90%) >SGD(85%) 로 Randomforest가 가장 높은 값을 보였다.
* accuary(정확도)로 비교해보면 Randomforest(98%) > Logistic Regression(94%) > SVM(92%) >Decision Tree(91%) > SGD(89%) 로  accuary에서도 Randomforest가 가장 높은 값을 보였다.

(3) 유방암 여부를 진단 하기 적합한 모델
* Random Forest가 가장 적절하다.